File: gtnh/builder/test/test_plan.lua
-- test_plan.lua — Tests for the plan format reader against golden plans.
-- Run with: node test_runner.js test_plan.lua
-- Requires the converter's golden plan files in converter/test/.

local plan = require("plan")

local passed = 0
local failed = 0

local function test(name, fn)
  local ok, err = pcall(fn)
  if ok then
    print("  PASS: " .. name)
    passed = passed + 1
  else
    print("  FAIL: " .. name .. " — " .. tostring(err))
    failed = failed + 1
  end
end

local function assert_eq(a, b, msg)
  if a ~= b then
    error(string.format("%s: expected %s, got %s", msg or "assert_eq", tostring(b), tostring(a)))
  end
end

-- Read a file from the converter's test directory.
-- The test runner provides a readFile function that uses Node.js fs.
local function readGoldenFile(filename)
  -- Try a few plausible paths (converter/test/, ../converter/test/, etc.)
  local paths = {
    "converter/test/" .. filename,
    "../converter/test/" .. filename,
    "test/" .. filename,
    "../../converter/test/" .. filename,
  }
  for _, p in ipairs(paths) do
    local data, err = readFile(p)
    if data then
      return data
    end
  end
  error("Could not find golden file: " .. filename)
end

print("=== Plan Reader Tests ===")

test("parse header from tree5.plan", function()
  local raw = readGoldenFile("tree5.plan")
  local p = plan.decode(raw)

  assert_eq(p.header.formatVersion, 1)
  assert_eq(p.header.width, 22, "width")
  assert_eq(p.header.height, 29, "height")
  assert_eq(p.header.length, 23, "length")
  assert_eq(p.header.planVersion, 0, "planVersion")
  assert_eq(p.name, "tree5", "name")
  assert_eq(p.palette[1].itemName, "log", "palette item")
  assert_eq(p.palette[1].damage, 1, "palette damage")
  assert_eq(p.palette[1].blockName, "log", "blockName")
  assert_eq(p.palette[1].meta, 1, "blockMeta")
  assert_eq(p.palette[1].flags, 1, "flags (orient)") -- 0x01 = FLAG_ORIENT
  assert_eq(#p.layers, 29, "layer count")
  assert_eq(#p.layerOffsets, 29, "offset count")
  -- Each layer should be W*L = 22*23 = 506 cells
  for y = 1, 29 do
    assert_eq(#p.layers[y], 506, "layer " .. y .. " size")
  end
end)

test("cellAt returns correct palette indices", function()
  local raw = readGoldenFile("tree5.plan")
  local p = plan.decode(raw)
  local W = p.header.width

  -- Layer 0: from the converter output, we know the layout.
  -- The manifest says 69 placed cells total across all layers.
  -- Let's just verify cellAt works by checking a few known cells.
  local layer0 = p.layers[1]

  -- Cell (0,0) should be palette index 0 (AIR) for tree5 (mostly air).
  local c = plan.cellAt(layer0, 0, 0, W)
  assert_eq(c, 0, "cell (0,0,0) is air")

  -- Find a non-air cell in layer 0.
  local found = false
  for x = 0, W - 1 do
    for z = 0, p.header.length - 1 do
      local c = plan.cellAt(layer0, x, z, W)
      if c > 1 then
        found = true
        assert(c <= #p.palette + 1, "palette index in range")
        break
      end
    end
    if found then break end
  end
  assert(found, "found at least one placed block")
end)

test("cellIndex math", function()
  assert_eq(plan.cellIndex(0, 0, 10), 0)
  assert_eq(plan.cellIndex(5, 0, 10), 5)
  assert_eq(plan.cellIndex(0, 1, 10), 10)
  assert_eq(plan.cellIndex(9, 3, 10), 9 + 3 * 10)
end)

test("palette constants", function()
  assert_eq(plan.PALETTE_AIR, 0)
  assert_eq(plan.PALETTE_SKIP, 1)
  assert_eq(plan.PALETTE_BASE, 2)
  assert_eq(plan.FLAG_ORIENT, 0x01)
  assert_eq(plan.FLAG_TILEENTITY, 0x02)
end)

test("bad magic rejected", function()
  local bad = "XXXX" .. string.rep("\0", 20)
  local ok, err = pcall(plan.decode, bad)
  assert(not ok, "should reject bad magic")
  assert(tostring(err):find("Bad magic"), "error should mention magic")
end)

test("CRC mismatch detected", function()
  local raw = readGoldenFile("tree5.plan")
  -- Corrupt a byte in the body.
  local corrupted = raw:sub(1, 25) .. "\xFF" .. raw:sub(27)
  local ok, err = pcall(plan.decode, corrupted)
  assert(not ok, "should reject CRC mismatch")
  assert(tostring(err):find("CRC"), "error should mention CRC")
end)

print(string.format("\n=== %d passed, %d failed ===", passed, failed))
os.exit(failed > 0 and 1 or 0)

File: converter/test/test_plan.js
#!/usr/bin/env node
'use strict';
// test_plan.js — unit tests for the plan format and converter.

const assert = require('assert');
const { encode, decode, PALETTE_AIR, PALETTE_SKIP, PALETTE_BASE, FLAG_ORIENT, FLAG_TILEENTITY } = require('../lib/plan.js');
const varint = require('../lib/varint.js');
const { crc32 } = require('../lib/crc32.js');

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (e) {
    console.error(`  ✗ ${name}`);
    console.error(`    ${e.message}`);
    failed++;
  }
}

console.log('=== Plan Format Tests ===\n');

// --- Varint tests ---
console.log('Varint:');
test('encode/decode small values', () => {
  for (const v of [0, 1, 127, 128, 255, 256, 16383, 16384, 1000000]) {
    const encoded = varint.encode(v);
    const decoded = varint.decode(encoded, 0);
    assert.strictEqual(decoded.value, v, `value ${v}`);
    assert.strictEqual(decoded.length, encoded.length, `length ${v}`);
  }
});

test('varint round-trip', () => {
  const buf = [];
  for (const v of [42, 0, 65535, 123456]) {
    varint.append(buf, v);
  }
  let pos = 0;
  for (const v of [42, 0, 65535, 123456]) {
    const r = varint.decode(buf, pos);
    assert.strictEqual(r.value, v);
    pos += r.length;
  }
});

// --- CRC32 tests ---
console.log('\nCRC32:');
test('CRC32 known value', () => {
  // "123456789" → 0xCBF43926
  const data = Buffer.from('123456789', 'ascii');
  assert.strictEqual(crc32(data), 0xCBF43926);
});

// --- Plan encode/decode round-trip ---
console.log('\nPlan encode/decode:');

function makeTestPlan() {
  const W = 3, H = 2, L = 2;
  const layers = [];
  // Layer 0: air (0), skip (1), log (2)
  layers.push([PALETTE_AIR, PALETTE_SKIP, 2, 2, 2, 2]);
  // Layer 1: log (2), leaves (3), air (0)
  layers.push([2, 3, 0, 0, 0, 0]);
  const palette = [
    { itemName: 'log', damage: 1, blockName: 'log', meta: 1, flags: 0 },
    { itemName: 'leaves', damage: 0, blockName: 'leaves', meta: 0, flags: FLAG_ORIENT },
  ];
  return { W, H, L, layers, palette, name: 'test', version: 12345 };
}

test('encode/decode round-trip', () => {
  const t = makeTestPlan();
  const buf = encode({ width: t.W, height: t.H, length: t.L, name: t.name, planVersion: t.version, layers: t.layers, palette: t.palette });
  const p = decode(buf);
  assert.strictEqual(p.header.width, t.W);
  assert.strictEqual(p.header.height, t.H);
  assert.strictEqual(p.header.length, t.L);
  assert.strictEqual(p.name, t.name);
  assert.strictEqual(p.header.planVersion, t.version);
  assert.strictEqual(p.palette.length, t.palette.length);
  assert.strictEqual(p.layers.length, t.H);
  // Verify cell values.
  assert.deepStrictEqual(p.layers[0], t.layers[0]);
  assert.deepStrictEqual(p.layers[1], t.layers[1]);
});

test('palette entry round-trip', () => {
  const t = makeTestPlan();
  const buf = encode({ width: t.W, height: t.H, length: t.L, name: t.name, planVersion: t.version, layers: t.layers, palette: t.palette });
  const p = decode(buf);
  assert.strictEqual(p.palette[0].itemName, 'log');
  assert.strictEqual(p.palette[0].damage, 1);
  assert.strictEqual(p.palette[0].blockName, 'log');
  assert.strictEqual(p.palette[0].meta, 1);
  assert.strictEqual(p.palette[0].flags, 0);
  assert.strictEqual(p.palette[1].itemName, 'leaves');
  assert.strictEqual(p.palette[1].flags, FLAG_ORIENT);
});

test('CRC verification', () => {
  const t = makeTestPlan();
  const buf = encode({ width: t.W, height: t.H, length: t.L, name: t.name, planVersion: t.version, layers: t.layers, palette: t.palette });
  // Decode verifies CRC automatically.
  decode(buf);
  // Corrupt a byte and verify CRC fails.
  const corrupted = Buffer.from(buf);
  corrupted[25] ^= 0xFF;
  assert.throws(() => decode(corrupted), /CRC mismatch/);
});

test('empty plan (all air)', () => {
  const W = 4, H = 3, L = 4;
  const layers = [];
  for (let y = 0; y < H; y++) {
    layers.push(new Array(W * L).fill(PALETTE_AIR));
  }
  const buf = encode({ width: W, height: H, length: L, name: 'empty', planVersion: 0, layers, palette: [] });
  const p = decode(buf);
  assert.strictEqual(p.header.width, W);
  assert.strictEqual(p.header.height, H);
  assert.strictEqual(p.header.length, L);
  assert.strictEqual(p.palette.length, 0);
  assert.strictEqual(p.layers.length, H);
});

test('large palette indices (varint multi-byte)', () => {
  const W = 2, H = 1, L = 2;
  const palette = [];
  // Create 300 palette entries to force varint multi-byte indices.
  for (let i = 0; i < 300; i++) {
    palette.push({ itemName: 'item' + i, damage: i, blockName: 'block' + i, meta: i, flags: 0 });
  }
  const layers = [new Array(W * L).fill(100 + PALETTE_BASE)]; // palette index 102
  const buf = encode({ width: W, height: H, length: L, name: 'big', planVersion: 0, layers, palette });
  const p = decode(buf);
  assert.strictEqual(p.palette.length, 300);
  assert.strictEqual(p.layers[0][0], 100 + PALETTE_BASE);
  assert.strictEqual(p.palette[100].itemName, 'item100');
});

console.log(`\n=== ${passed} passed, ${failed} failed ===`);
process.exit(failed > 0 ? 1 : 0);
